### ISSUE TITLE
Missing types when using project references and single run inference

### ISSUE THREAD

**Issue by gcangussu** [typescript-eslint/typescript-eslint#3851]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

We need a few files to reproduce this:

- a TS project
  - with a [custom path resolution to itself](https://github.com/gcangussu/typescript-eslint-bug-repro/blob/14d668bd4fe7387160e8e879830b9f2de02aef09/packages/bar/tsconfig.json#L5)
  - and [exporting something](https://github.com/gcangussu/typescript-eslint-bug-repro/blob/14d668bd4fe7387160e8e879830b9f2de02aef09/packages/bar/src/public/utils.ts#L3) that depends on a [imported symbol that is imported using the custom path
](https://github.com/gcangussu/typescript-eslint-bug-repro/blob/14d668bd4fe7387160e8e879830b9f2de02aef09/packages/bar/src/public/utils.ts#L1)
- other TS project[ using the above export](https://github.com/gcangussu/typescript-eslint-bug-repro/blob/14d668bd4fe7387160e8e879830b9f2de02aef09/packages/foo/src/app.ts#L2) of the first TS project

Reproduction is available here: https://github.com/gcangussu/typescript-eslint-bug-repro

**Expected Result**

Tests passing

```
$ npm --silent test && echo $?
0
```

We can see that it works with single run inference disabled with

```
npm run test-without-single-run-inference
```

**Actual Result**

```
npm --silent test && echo $?

/Users/gabriel/Repos/typescript-eslint-bug-repro/packages/foo/src/app.ts
   9:7   error  Unsafe assignment of an `any` value                @typescript-eslint/no-unsafe-assignment
  10:30  error  Invalid type "any" of template literal expression  @typescript-eslint/restrict-template-expressions
  10:30  error  Unsafe member access .toFixed on an `any` value    @typescript-eslint/no-unsafe-member-access
  10:30  error  Unsafe call of an `any` typed value                @typescript-eslint/no-unsafe-call

✖ 4 problems (4 errors, 0 warnings)

```

**Additional Info**

It seems that with single run the project (tsconfig.json) is not being correctly taken into consideration. Checking the [output of `packages/bar/src/public/utils.ts`](https://github.com/gcangussu/typescript-eslint-bug-repro/blob/14d668bd4fe7387160e8e879830b9f2de02aef09/build/packages/bar/src/public/utils.d.ts#L1) shows that it is correctly typed. However, when it is imported from another package (foo), ts-eslint fails to find its type. Although, it works fine when [imported from within the same package](https://github.com/gcangussu/typescript-eslint-bug-repro/blob/14d668bd4fe7387160e8e879830b9f2de02aef09/packages/bar/src/internal/app.ts#L2).

I've kept only the rule `@typescript-eslint/no-unsafe-assignment` and run `npx eslint --ext=.ts --debug packages/foo/src/app.ts`.

<details>
  <summary>Output, click to expand</summary>
  
```
2021-09-06T20:51:40.919Z eslint:cli CLI args: [ '--ext=.ts', '--debug', 'packages/foo/src/app.ts' ]
2021-09-06T20:51:40.921Z eslint:cli Running on files
2021-09-06T20:51:40.930Z eslintrc:config-array-factory Loading .eslintignore file: /Users/gabriel/Repos/typescript-eslint-bug-repro/.eslintignore
2021-09-06T20:51:40.931Z eslintrc:ignore-pattern Create with: [ IgnorePattern { patterns: [ '/**/node_modules/*' ], basePath: '/Users/gabriel/Repos/typescript-eslint-bug-repro', loose: false } ]
2021-09-06T20:51:40.933Z eslintrc:ignore-pattern   processed: { basePath: '/Users/gabriel/Repos/typescript-eslint-bug-repro', patterns: [ '/**/node_modules/*' ] }
2021-09-06T20:51:40.933Z eslintrc:ignore-pattern Create with: [ IgnorePattern { patterns: [ '/**/node_modules/*' ], basePath: '/Users/gabriel/Repos/typescript-eslint-bug-repro', loose: false } ]
2021-09-06T20:51:40.933Z eslintrc:ignore-pattern   processed: { basePath: '/Users/gabriel/Repos/typescript-eslint-bug-repro', patterns: [ '/**/node_modules/*' ] }
2021-09-06T20:51:40.934Z eslint:file-enumerator Start to iterate files: [ 'packages/foo/src/app.ts' ]
2021-09-06T20:51:40.934Z eslint:file-enumerator File: /Users/gabriel/Repos/typescript-eslint-bug-repro/packages/foo/src/app.ts
2021-09-06T20:51:40.934Z eslintrc:cascading-config-array-factory Load config files for /Users/gabriel/Repos/typescript-eslint-bug-repro/packages/foo/src.
2021-09-06T20:51:40.934Z eslintrc:cascading-config-array-factory No cache found: /Users/gabriel/Repos/typescript-eslint-bug-repro/packages/foo/src.
2021-09-06T20:51:40.934Z eslintrc:config-array-factory Config file not found on /Users/gabriel/Repos/typescript-eslint-bug-repro/packages/foo/src
2021-09-06T20:51:40.935Z eslintrc:cascading-config-array-factory No cache found: /Users/gabriel/Repos/typescript-eslint-bug-repro/packages/foo.
2021-09-06T20:51:40.935Z eslintrc:config-array-factory Config file not found on /Users/gabriel/Repos/typescript-eslint-bug-repro/packages/foo
2021-09-06T20:51:40.935Z eslintrc:cascading-config-array-factory No cache found: /Users/gabriel/Repos/typescript-eslint-bug-repro/packages.
2021-09-06T20:51:40.935Z eslintrc:config-array-factory Config file not found on /Users/gabriel/Repos/typescript-eslint-bug-repro/packages
2021-09-06T20:51:40.935Z eslintrc:cascading-config-array-factory No cache found: /Users/gabriel/Repos/typescript-eslint-bug-repro.
2021-09-06T20:51:40.935Z eslintrc:config-array-factory Loading JSON config file: /Users/gabriel/Repos/typescript-eslint-bug-repro/.eslintrc.json
2021-09-06T20:51:40.935Z eslintrc:config-array-factory Config file found: /Users/gabriel/Repos/typescript-eslint-bug-repro/.eslintrc.json
2021-09-06T20:51:40.936Z eslintrc:config-array-factory Loading parser "@typescript-eslint/parser" from /Users/gabriel/Repos/typescript-eslint-bug-repro/.eslintrc.json
2021-09-06T20:51:40.936Z eslintrc:config-array-factory Loaded: @typescript-eslint/parser@4.30.0 (/Users/gabriel/Repos/typescript-eslint-bug-repro/node_modules/@typescript-eslint/parser/dist/index.js)
2021-09-06T20:51:41.136Z eslintrc:config-array-factory Loading plugin "@typescript-eslint" from /Users/gabriel/Repos/typescript-eslint-bug-repro/.eslintrc.json
2021-09-06T20:51:41.137Z eslintrc:config-array-factory Loaded: @typescript-eslint/eslint-plugin@4.30.0 (/Users/gabriel/Repos/typescript-eslint-bug-repro/node_modules/@typescript-eslint/eslint-plugin/dist/index.js)
2021-09-06T20:51:41.274Z eslintrc:config-array-factory Plugin /Users/gabriel/Repos/typescript-eslint-bug-repro/node_modules/@typescript-eslint/eslint-plugin/dist/index.js loaded in: 137ms
2021-09-06T20:51:41.274Z eslintrc:cascading-config-array-factory Stop traversing because of 'root:true'.
2021-09-06T20:51:41.276Z eslintrc:cascading-config-array-factory Configuration was determined: ConfigArray(3) [ { type: 'config', name: 'DefaultIgnorePattern', filePath: '', criteria: null, env: undefined, globals: undefined, ignorePattern: IgnorePattern { patterns: [Array], basePath: '/Users/gabriel/Repos/typescript-eslint-bug-repro', loose: false }, noInlineConfig: undefined, parser: undefined, parserOptions: undefined, plugins: undefined, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: undefined, settings: undefined }, { type: 'config', name: '.eslintrc.json', filePath: '/Users/gabriel/Repos/typescript-eslint-bug-repro/.eslintrc.json', criteria: null, env: undefined, globals: undefined, ignorePattern: undefined, noInlineConfig: undefined, parser: { error: null, filePath: '/Users/gabriel/Repos/typescript-eslint-bug-repro/node_modules/@typescript-eslint/parser/dist/index.js', id: '@typescript-eslint/parser', importerName: '.eslintrc.json', importerPath: '/Users/gabriel/Repos/typescript-eslint-bug-repro/.eslintrc.json' }, parserOptions: { allowAutomaticSingleRunInference: true, project: [Array] }, plugins: { '@typescript-eslint': [Object] }, processor: undefined, reportUnusedDisableDirectives: undefined, root: true, rules: { '@typescript-eslint/no-unsafe-assignment': 'error' }, settings: undefined }, { type: 'ignore', name: '.eslintignore', filePath: '/Users/gabriel/Repos/typescript-eslint-bug-repro/.eslintignore', criteria: null, env: undefined, globals: undefined, ignorePattern: IgnorePattern { patterns: [Array], basePath: '/Users/gabriel/Repos/typescript-eslint-bug-repro', loose: true }, noInlineConfig: undefined, parser: undefined, parserOptions: undefined, plugins: undefined, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: undefined, settings: undefined } ] on /Users/gabriel/Repos/typescript-eslint-bug-repro/packages/foo/src
2021-09-06T20:51:41.276Z eslintrc:ignore-pattern Create with: [ IgnorePattern { patterns: [ '/**/node_modules/*' ], basePath: '/Users/gabriel/Repos/typescript-eslint-bug-repro', loose: false }, IgnorePattern { patterns: [ 'build/' ], basePath: '/Users/gabriel/Repos/typescript-eslint-bug-repro', loose: true } ]
2021-09-06T20:51:41.276Z eslintrc:ignore-pattern   processed: { basePath: '/Users/gabriel/Repos/typescript-eslint-bug-repro', patterns: [ '/**/node_modules/*', 'build/' ] }
2021-09-06T20:51:41.277Z eslintrc:ignore-pattern Check {
  filePath: '/Users/gabriel/Repos/typescript-eslint-bug-repro/packages/foo/src/app.ts',
  dot: false,
  relativePath: 'packages/foo/src/app.ts',
  result: false
}
2021-09-06T20:51:41.277Z eslint:cli-engine Lint /Users/gabriel/Repos/typescript-eslint-bug-repro/packages/foo/src/app.ts
2021-09-06T20:51:41.277Z eslint:linter Linting code for /Users/gabriel/Repos/typescript-eslint-bug-repro/packages/foo/src/app.ts (pass 1)
2021-09-06T20:51:41.277Z eslint:linter Verify
2021-09-06T20:51:41.277Z eslint:linter With ConfigArray: /Users/gabriel/Repos/typescript-eslint-bug-repro/packages/foo/src/app.ts
2021-09-06T20:51:42.117Z eslint:linter Generating fixed text for /Users/gabriel/Repos/typescript-eslint-bug-repro/packages/foo/src/app.ts (pass 1)
2021-09-06T20:51:42.117Z eslint:source-code-fixer Applying fixes
2021-09-06T20:51:42.117Z eslint:source-code-fixer shouldFix parameter was false, not attempting fixes
2021-09-06T20:51:42.118Z eslint:file-enumerator Complete iterating files: ["packages/foo/src/app.ts"]
2021-09-06T20:51:42.118Z eslint:cli-engine Linting complete in: 1184ms

/Users/gabriel/Repos/typescript-eslint-bug-repro/packages/foo/src/app.ts
  9:7  error  Unsafe assignment of an `any` value  @typescript-eslint/no-unsafe-assignment

✖ 1 problem (1 error, 0 warnings)

```
</details>

Relates to #3528

**Versions**

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/typescript-estree` | `4.30.0` |
| `@typescript-eslint/eslint-plugin` | `4.30.0` |
| `@typescript-eslint/parser` | `4.30.0` |
| `TypeScript`                           | `4.4.2` |
| `node`                                 | `14.17.6` |

**Comment by bradzacher** [comment]
We don't consume project references from this project - we never have - #2094.
Which is why I'm ***very*** surprised that your config lints.

In a nutshell - when in single-run mode we do a different 

It looks like there's something internal to TS which means that:
`ts.getParsedCommandLineOfConfigFile(...).fileNames` returns
```
[
  '~/typescript-eslint-bug-repro-main/packages/bar/src/internal/utils.ts',
  '~/typescript-eslint-bug-repro-main/packages/bar/src/public/utils.ts',
  '~/typescript-eslint-bug-repro-main/packages/foo/src/app.ts'
]
```

In comparison the program created by `ts.createWatchProgram` instead returns
```
[
  '~/typescript-eslint-bug-repro-main/build/packages/bar/src/internal/utils.d.ts',
  '~/typescript-eslint-bug-repro-main/build/packages/bar/src/public/utils.d.ts',
  '~/typescript-eslint-bug-repro-main/packages/foo/src/app.ts'
]
```

I'm not sure why they behave differently.

As far as I can tell - this is likely related to your `paths` config. If it were due to the project references or single run inference then our linting within this project (which uses both) would be broken.

I won't have any bandwidth to look into this more sorry - happy to provide pointers if you'd like to dig in more.

**Comment by gcangussu** [comment]
Thanks for taking a look at it. And sure, If you could provide some _points of interest_ on the code that would be great for me to investigate.

**Comment by bradzacher** [comment]
1) This function is the entrypoint into the parser from ESLint:
  - https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/src/parser.ts#L511
2) This code is used by the "single run inference" to construct a program
  - https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/src/create-program/useProvidedPrograms.ts#L50-L87
  - Which is called from here: https://github.com/typescript-eslint/typescript-eslint/blob/ebb33ed8bc29f69ca2a657ec5b31857c0aeb4b56/packages/typescript-estree/src/parser.ts#L555-L574
3) This code is used by the non-single-run codepath to construct the "watch program"
  - https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/src/create-program/createWatchProgram.ts#L142
  - Which is called from here: https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/src/create-program/createProjectProgram.ts#L26


in a nutshell - the code follows this structure:
1) `parseAndGenerateServices` ((1) above)
2) if is single run, create an iterable for each of the `tsconfig`s provided ((2) above)
3) else create a "watch program" for each of the `tsconfig`s provided  ((3) above)

It looks like for some reason TS's `ts.getParsedCommandLineOfConfigFile` API (used in (2) above) resolves the list of files without considering project references - leading to files that exist outside the program (causing errors).

You might need to dig into the TS codebase a little to figure out if we're calling the API incorrectly, or if we're calling the `ts.createProgram` API incorrectly, or if this is just something weird in TS we'd need to work around or report upstream.

**Comment by aaronadamsCA** [comment]
When I add `allowAutomaticSingleRunInference: true` to my parser options, `@typescript-eslint/strict-boolean-expressions` flags this error on every linted file:

``This rule requires the `strictNullChecks` compiler option to be turned on to function correctly``

I have `"strict": true` in my root `tsconfig.json`, which is extended by each project's own `tsconfig.json`, and it all works fine without single run inference.

Sharing in case this points to a different root cause.

**Comment by JoshuaKGoldberg** [comment]
Sadly, I think this is blocked on us being able to know whether ESLint is in fixer mode or not. I think that makes this issue:

* A duplicate of #2094 
* Blocked on https://github.com/eslint/rfcs/pull/102

🥲, as always.

**Comment by JoshuaKGoldberg** [comment]
Well, the ESLint RFC was declined and no action has happened on that front since then (😞). But we would greatly benefit from being able to enable `allowAutomaticSingleRunInference` by default.

A hunch: I bet we can do something "clever" internally to turn off `allowAutomaticSingleRunInference` if linting re-runs in a manner suspiciously like `--fix`. Thoughts @bradzacher?

Marking as `accepting prs` in the meantime & adding to the v7 milestone.

**Comment by JoshuaKGoldberg** [comment]
Ok! Following up on this: I now more strongly believe the project service (`EXPERIMENTAL_useProjectService` / `useProjectService`) is the solution here. I updated the repro repo's dependency versions to latest and enabled it in the `.eslintrc.json`. Running `npm run test` showed no erroneous lint complaints. Whoo!

I would love to know why the watch vs. non-watch programs go down different code paths. But given that the new project service resolves the issue and we don't actually support project references otherwise, I don't think we need to take a strong look at this. Closing as resolved.

```diff
  "parserOptions": {
    "allowAutomaticSingleRunInference": true,
-   "project": ["./packages/*/tsconfig.json"]
+   "EXPERIMENTAL_useProjectService": true
  },
```

Someone please yell at me if we actually do have a need to investigate more deeply! 💜
